inginious/agent/docker_agent/__init__.py

```diff
@@ -437,13 +443,22 @@ async def create_student_container(self, parent_info, socket_id, environment_nam
             environment = self._containers[environment_type][environment_name]["id"]
             runtime = self._containers[environment_type][environment_name]["runtime"]
 
+            ports_needed = [22] if ssh else []
+            ports = {}
+            for p in ports_needed:
+                if len(self._external_ports) == 0:
+                    self._logger.warning("User asked for a port but no one are available")
+                    raise CannotCreateJobException('No ports are available right now. Please retry later.')
```

Comment:
Caution: you should ensure that the ports that were already in `ports` are going back to `self._external_ports`.

Reply:
Indeed, actually I had in mind cases with a single port needed so the specific case you are talking about was not possible (since "ports" would be empty)... But indeed since the student_container may ask for multiple ports, some could be given to him before realising there are no enough ports available !
Well seen, I would not have remarked this myself.